core/silkworm/execution/evm.cpp

```diff
@@ -154,62 +154,63 @@ evmc::result EVM::create(const evmc_message& message) noexcept {
 evmc::result EVM::call(const evmc_message& message) noexcept {
     evmc::result res{EVMC_SUCCESS, message.gas, nullptr, 0};
 
-    auto value{intx::be::load<intx::uint256>(message.value)};
+    const auto value{intx::be::load<intx::uint256>(message.value)};
     if (message.kind != EVMC_DELEGATECALL && state_.get_balance(message.sender) < value) {
         res.status_code = EVMC_INSUFFICIENT_BALANCE;
         return res;
     }
 
-    const bool precompiled{is_precompiled(message.destination)};
+    // See Section 8 "Message Call" of the Yellow Paper for the difference between code & recipient.
+    // destination in evmc_message can mean either code or recipient, depending on the context.
+    const evmc_address code_address{message.destination};
+    const evmc_address recipient_address{recipient_of_call_message(message)};
+
+    const bool precompiled{is_precompiled(code_address)};
     const evmc_revision rev{revision()};
 
     // https://eips.ethereum.org/EIPS/eip-161
-    if (value == 0 && rev >= EVMC_SPURIOUS_DRAGON && !state_.exists(message.destination) && !precompiled) {
+    if (value == 0 && rev >= EVMC_SPURIOUS_DRAGON && !precompiled && !state_.exists(code_address)) {
         return res;
     }
 
-    auto snapshot{state_.take_snapshot()};
+    const auto snapshot{state_.take_snapshot()};
 
     if (message.kind == EVMC_CALL) {
         if (message.flags & EVMC_STATIC) {
             // Match geth logic
             // https://github.com/ethereum/go-ethereum/blob/v1.9.25/core/vm/evm.go#L391
-            state_.touch(message.destination);
+            state_.touch(recipient_address);
```

Comment:
This seems to be fixing a consensus bug. Are there any tests failing?

Reply:
> This seems to be fixing a consensus bug. Are there any tests failing?

Not really because this code is executed only when `message.kind == EVMC_CALL`